### PR TITLE
Unboxed returns (multiple return values)

### PR DESCRIPTION
**Pull request by Ekdohibs**
This is a rebase of #1170, together with an attribute `[@unboxed]` that can be added on the definition of a function to decide to have an unboxed result. It does not trigger automatically, as that would require thinking about a probably complex strategy to avoid reallocations (leading to two test failures).

### CODE REVIEW

**Comment by Gbury** [comment]
I've found a bug ! Compiling and executing the following code:

```ocaml

let[@inline never][@unboxed] f x =
  ( (* each line has 20 elements *)
    x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x,
    x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x,
    x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x,
    x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x,
    x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x, x
  )

let () =
  let (
    _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _,
    _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _,
    _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _,
    _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _,
    _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, y, _, _, _
  )
    = f 5
  in
  print_int y;
  print_newline ()
```

Results in the output `0` instead of `5`.

The cause seems to be that after the call, the stack pointer is moved before moving the result into a register.

It might be useful to add a few tests exercising unboxed result somewhere.

**Comment by Gbury** [comment]
the bug above is fixed by https://github.com/ocaml-flambda/flambda-backend/pull/1248/commits/f50cb88ca03aae3a9553ee5b95a8bb0192c21b84

File: backend/spill.ml [hunk 1]
```diff
@@ -617,6 +614,10 @@ let rec spill :
   | Iraise _ ->
       k i env.at_raise
 
+let reset_cache () =
+  reload_cache := Numbers.Int.Map.empty;
+  spill_cache := Numbers.Int.Map.empty
```
[the remark this answers is not in the document]
Comment (lthls) [reply backend/spill.ml:619]: Yes, there was. In the presence of duplicate labels (caused by mixing `Lambda.next_raise_count` and `Cmm.new_label`), not resetting the spill cache properly created even more issues, so I added the patch.
It's less necessary now, mostly just preventing memory leaks, but if we don't include this patch in this pull request I'm likely to forget about it completely.

If you want to submit it separately, feel free to do so.

File: middle_end/flambda2/identifiers/continuation.ml [hunk 2]
```diff
@@ -114,9 +114,9 @@ let reset () = initialise ()
 
 let create ?sort ?name () : t =
   let sort = Option.value sort ~default:Sort.Normal_or_exn in
-  let name = Option.value name ~default:"k" in
-  let compilation_unit = Compilation_unit.get_current_exn () in
   let name_stamp = next_stamp () in
+  let name = Option.value name ~default:(Format.asprintf "k_%d" name_stamp) in
```
[the remark this answers is not in the document]
Comment (Gbury) [reply middle_end/flambda2/identifiers/continuation.ml:118]: Indeed, I am planning to make a PR to add a debug option for flambda, and one of its effects would be to give unique names to continuations.

[the remark this answers is not in the document]
Comment (Gbury) [reply middle_end/flambda2/identifiers/continuation.ml:118]: So I think it's better to remove this code from the PR.

File: backend/spill.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gbury) [reply backend/spill.ml:619]: I'm fine with it being part of this patch, I was just wondering what it changed, so your answer is just what I wanted.

**Comment by Ekdohibs** [comment]
Subsumed by #1271.

**Pull request closed without merging**